### PR TITLE
Allow generate record with g_boxed_copy and g_boxed_free

### PR DESCRIPTION
**Pull request by EPashkin**
Alternative of https://github.com/gtk-rs/gir/pull/542
Add `record_boxed` config for records.

### CODE REVIEW

File: src/codegen/general.rs [hunk 1]
```diff
@@ -149,6 +149,48 @@ pub fn define_boxed_type(
...
+    try!(writeln!(w, ""));
+    try!(writeln!(w, "glib_wrapper! {{"));
+    try!(writeln!(
+        w,
+        "\tpub struct {}(Boxed<ffi::{}>);",
+        type_name,
+        glib_name
+    ));
+    try!(writeln!(w, ""));
+    try!(writeln!(w, "\tmatch fn {{"));
```
Comment (GuillaumeGomez) [inline src/codegen/general.rs]: We have tabulations now? :O

Comment (EPashkin) [reply src/codegen/general.rs]: And long long time ;)

Comment (GuillaumeGomez) [reply src/codegen/general.rs]: Oh damn. I'm so late!

Comment (EPashkin) [reply src/codegen/general.rs]: It replaced with 4 spaces by https://github.com/gtk-rs/gir/blob/master/src/writer/untabber.rs

Comment (GuillaumeGomez) [reply src/codegen/general.rs]: Just saw it a few minutes ago (I was very surprised). Getting old, my memory can't follow up...

File: README.md [hunk 2]
```diff
@@ -203,6 +203,18 @@ version = "3.12"
     version = "3.18"
 ```
 
+Some records have non-standard `copy` or `free` functions or even do not have it.
+You can define these records with functions for boxed types.
```
Comment (MathieuDuponchelle) [inline README.md]: I think the name does not really reflect the intent here, a record might be defined as a boxed type but would not need to use g_boxed functions, how about `use_boxed_functions` maybe?

As for the documentation comment I would then go for:

```
gir auto-detects copy and free functions by looking up `type_name_copy` and `type_name_free` functions.
If the relevant functions are not named that way, and the type is defined with `G_DEFINE_BOXED_TYPE`,
`use_boxed_functions=true` can be used to call `g_boxed_copy` and `g_boxed_free` instead.

> In that case, it is required to declare the `get_type` function in the headers.
```

That is a bit verbose but would make me less confused :)

Comment (EPashkin) [reply README.md]: Thanks, I make change

**Comment by EPashkin** [comment]
Now gir generate `AutoBoxed` record as fall-back.
`RecordType::Direct` commented until we have type to generate for it

**Comment by sdroege** [comment]
Can you put a diff somewhere for the effect this has on one of the crates?

**Comment by EPashkin** [comment]
No changes in gtk-rs crates by default as old fall-back was panic.
Changes for `use_boxed_functions = true`
```diff
--- a/src/auto/paper_size.rs
+++ b/src/auto/paper_size.rs
@@ -15,8 +16,8 @@ glib_wrapper! {
     pub struct PaperSize(Boxed<ffi::GtkPaperSize>);
 
     match fn {
-        copy => |ptr| ffi::gtk_paper_size_copy(mut_override(ptr)),
-        free => |ptr| ffi::gtk_paper_size_free(ptr),
+        copy => |ptr| gobject_ffi::g_boxed_copy(ffi::gtk_paper_size_get_type(), ptr as *mut _) as *mut ffi::GtkPaperSize,
+        free => |ptr| gobject_ffi::g_boxed_free(ffi::gtk_paper_size_get_type(), ptr as *mut _),
         get_type => || ffi::gtk_paper_size_get_type(),
     }
 }
```

**Comment by sdroege** [comment]
Seems good to me, also the changes in GIR itself.

**Pull request merged**